docs/hinsage.txt

```diff
@@ -0,0 +1,450 @@
+Heterogeneous GraphSAGE (HinSAGE)
+===================================
+
+This document outlines the viability and potential methodology to
+generalise the GraphSAGE algorithm [1] for
+heterogeneous graphs i.e. graphs containing many different node and edge
+types.
+
+Feature updates for homogeneous graphs
+--------------------------------------
+
+The feature update rule for homogeneous graphs is, for mean aggregator:
+
+1. Aggregation of features from the neighbours of node :math:`v`:
+
+   :math:`{h^{k}}_{N(v)} = \frac{1}{|N(v)|}D_{p}\lbrack{h_{u}}^{k - 1}\rbrack`
+
+2. Forward pass through layer :math:`k`:
+
+   If concat=True:
+
+   :math:`{h_{v}}^{k} = \sigma\left( concat\lbrack{W^{k}}_{\text{self}}D_{p}\lbrack{h_{v}}^{k - 1}\rbrack,\ {W^{k}}_{\text{neigh}}{{h}^{k}}_{N(v)}\rbrack + b^{k} \right)`
+
+   If concat=False:
+
+   :math:`{h_{v}}^{k} = \sigma\left( {W^{k}}_{\text{self}}D_{p}\lbrack{h_{v}}^{k - 1}\rbrack + \ {W^{k}}_{\text{neigh}}{h^{k}}_{N(v)} + b^{k} \right)`
+
+Where:
+
+-  :math:`{W^{k}}_{\text{self}}`\ and :math:`{W^{k}}_{\text{neigh}}` (both of size :math:`d_{k}/2 \times d_{k - 1}` if concat=True, or of size :math:`{d_{k}}^{} \times d_{k - 1}` if concat=False) are trainable parameters (shared for all nodes :math:`v`),
+
+-  :math:`b^{k}` is an optional bias,
+
+-  :math:`d_{k}` is node feature dimensionality at layer :math:`k`,
+
+-  :math:`\sigma` is the nonlinear activation,
+
+-  :math:`N(v)` is the neighbourhood of node :math:`v`
+
+-  :math:`D_{\text{p\ }}\lbrack \cdot \rbrack`\ is a random dropout with probability :math:`p` applied to its argument vector.
+
+The number of trainable parameters in layer :math:`k` for the mean
+aggregator is
+
+   :math:`{{2d}_{k}}/2 \times d_{k - 1} + d_{k} = d_{k}d_{k - 1} + d_{k}` if concat=True, or
```

Comment:
the use of `/` here instead of `\frac` makes it hard to read. Similar issues with other equations below.

```suggestion
   :math:`\frac{{{2d}_{k}}}{2} \times d_{k - 1} + d_{k} = d_{k}d_{k - 1} + d_{k}` if concat=True, or
```